official/resnet/resnet.py

```diff
@@ -169,48 +175,122 @@ def building_block(inputs, filters, training, projection_shortcut, strides,
 
   Args:
     inputs: A tensor of size [batch, channels, height_in, width_in] or
       [batch, height_in, width_in, channels] depending on data_format.
-    filters: The number of filters for the first two convolutions. Note
-      that the third and final convolution will use 4 times as many filters.
+    filters: The number of filters for the convolutions.
     training: A Boolean for whether the model is in training or inference
       mode. Needed for batch normalization.
     projection_shortcut: The function to use for projection shortcuts
       (typically a 1x1 convolution when downsampling the input).
     strides: The block's stride. If greater than 1, this block will ultimately
       downsample the input.
-    data_format: The input format ('channels_last' or 'channels_first').
+    data_format: The input format ('channels_last' or 'channels_first')..
 
   Returns:
     The output tensor of the block.
   """
   shortcut = inputs
   inputs = batch_norm_relu(inputs, training, data_format)
 
+  # The projection shortcut should come after the first batch norm and ReLU
+  # since it performs a 1x1 convolution.
+  if projection_shortcut is not None:
+    shortcut = projection_shortcut(inputs)
+
+  inputs = conv2d_fixed_padding(
+      inputs=inputs, filters=filters, kernel_size=3, strides=strides,
+      data_format=data_format)
+
+  inputs = batch_norm_relu(inputs, training, data_format)
+  inputs = conv2d_fixed_padding(
+      inputs=inputs, filters=filters, kernel_size=3, strides=1,
+      data_format=data_format)
+
+  return inputs + shortcut
+
+
+def _bottleneck_block_v1(inputs, filters, training, projection_shortcut,
+    strides, data_format):
+  """
+  Similar to _building_block_v1(), except using the "bottleneck" blocks
+  described in:
+    Convolution then batch normalization then ReLU as described by:
+      Deep Residual Learning for Image Recognition
+      https://arxiv.org/pdf/1512.03385.pdf
+      by Kaiming He, Xiangyu Zhang, Shaoqing Ren, and Jian Sun, Dec 2015.
+  """
+  shortcut = inputs
+
+  if projection_shortcut is not None:
+    shortcut = projection_shortcut(inputs)
+    shortcut = batch_norm_relu(shortcut, training, data_format)
+
+  inputs = conv2d_fixed_padding(
+      inputs=inputs, filters=filters, kernel_size=1, strides=1,
+      data_format=data_format)
+  inputs = batch_norm_relu(inputs, training, data_format)
+
+  inputs = conv2d_fixed_padding(
+      inputs=inputs, filters=filters, kernel_size=3, strides=strides,
+      data_format=data_format)
+  inputs = batch_norm_relu(inputs, training, data_format)
+
+  inputs = conv2d_fixed_padding(
+      inputs=inputs, filters=4 * filters, kernel_size=1, strides=1,
+      data_format=data_format)
+  inputs = batch_norm_relu(inputs, training, data_format,
+                           intermediate_addition=shortcut)
+
+  return inputs
+
+
+def _bottleneck_block_v2(inputs, filters, training, projection_shortcut,
```

Comment:
Can we actually combine the v1 and v2 functions into a single function? There is a lot of repeated code here, and I think it should also be simpler after the `batch_norm_relu` fix I suggested above.

Reply:
I agree, it doesn't seem ideal. As an exercise I put together an ultra concise function that could do v1 and v2 as well as regular and bottleneck.

```
_BOTTLENECK_LAYERS = [(1, 1, False, False), (3, 1, False, True), (1, 4, True, False)]
_REGULAR_LAYERS = [(3, 1, False, True), (3, 1, True, False)]
def _build_block(inputs, filters, training, projection_shortcut, strides,
    data_format, version, bottleneck):
  """
  Docstring removed for easier viewing on GitHub.
  """
  shortcut = inputs
  if version == 2:
    inputs = batch_norm_relu(inputs, training, data_format)

  if projection_shortcut is not None:
    shortcut = projection_shortcut(inputs)

  layers = _BOTTLENECK_LAYERS if bottleneck else _REGULAR_LAYERS
  for kernel_size, filter_num_mult, is_final, use_strides in layers:
    inputs = conv2d_fixed_padding(
        inputs=inputs, filters=filters, kernel_size=kernel_size,
        strides=strides if use_strides else 1,
        data_format=data_format)

    if version == 1:
      inputs = batch_norm_relu(inputs, training, data_format,
                       intermediate_addition=shortcut if is_final else None)
    else:  # version in (1, 2) check is performed by the resnet model
      if is_final:
        inputs += shortcut
      else:
        inputs = batch_norm_relu(inputs, training, data_format)
  return inputs
```

Even though this is < 50 lines (including the full docstring) compared to the ~170 being used now,  it's not easy to read. I think this is one of the rare cases where some code duplication may be good simply to improve readability. (That doesn't necessarily mean what is there now is optimal.) @karmel thoughts?

Reply:
I agree, combining both v1 + v2 and regular + bottleneck is likely too much. I think it's fine to combine just v1 + v2.

Reply:
Questions one might use to decide:

1. Are people more likely to want to read through the code for one version, or to compare the two versions? I feel like the answer to this is probably "read through one version," as most users want to run one version of ResNet, rather than use this as the place to learn about the difference between the two. That would imply the current (separated) version is preferable, because it's easier to parse (and extract) each individual model.
2. Are developers more likely to make mistakes with duplicated code or branching code? Both are tricky. Duplicated might be slightly more error-prone, but branching code is pretty bad too.
3. What does the rest of the code do-- duplicate or branch? We use  some of each pattern. We duplicate for imagenet versus cifar, but branch in some of the internals in deciding how many strides, etc. etc. We seem to err on the side of duplication, though.

There doesn't seem to be an obvious answer, but I would say I lean slightly in favor of duplication in the name of clarity given the above. Assuming this is correct WRT what the ops are supposed to be, we can go with this, and reassess if we find that we are changing the code a lot and running into maintenance problems.

Reply:
OK, I see why you both feel duplication results in less complication for the reader. My main concern here is that our code is getting very long (resnet.py is ~800 lines as of this PR), and that by adding lots of different features we are also adding a lot of extra code paths which can dilute the value of our code to the reader.

I agree now that combining these functions together with extra logic probably doesn't improve the situation much, but my main point is that keeping our code concise is something we should keep in mind.